File: packages/sdk-py/pylva/core/budget_accumulator.py
"""B2a — SDK-side budget accumulator (Python parity with TS).

Per-process dict keyed on ``{rule_id}:{scope_token}:{period_start}`` with LRU
eviction at 50,000 entries. Threading primitives (Lock + Timer) keep the
accumulator safe under concurrent wrapper calls without forcing asyncio.
"""

from __future__ import annotations

import threading
import time
from collections import OrderedDict
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Literal

import httpx

from .config import get_config, get_config_generation

LRU_MAX = 50_000
SYNC_INTERVAL_SEC = 5 * 60
# The backend accepts at most 500 entries per /budget/sync request. Split large
# per-customer snapshots so one builder crossing that limit does not disable
# reconciliation for every accumulator entry.
SYNC_BATCH_SIZE = 500

Scope = Literal["per_customer", "pooled"]
Period = Literal["hour", "day", "week", "month"]


@dataclass
class AccumulatorEntry:
    total_usd: float = 0.0
    event_count: int = 0
    last_touched: float = 0.0
    exceeded_source: Literal["backend_ingest_flag"] | None = None


@dataclass(frozen=True)
class AccumulatorKey:
    rule_id: str
    scope: Scope
    customer_id: str | None
    period_start: str


# OrderedDict preserves insertion order for LRU semantics.
_accum: OrderedDict[str, AccumulatorEntry] = OrderedDict()
_lock = threading.Lock()
_timer_lock = threading.Lock()
_sync_timer: threading.Timer | None = None
_cache_epoch = 0
_accepted_config_generation: int | None = None
_lru_warned_at: float = 0.0
_sync_missing_period_start_warned: bool = False
_sync_ambiguous_period_start_warned: bool = False


def _canonical_period_start(period_start: str) -> str:
    """Return JS Date.toISOString()-style UTC milliseconds, or the original string.

    Period strings are part of the accumulator key, so Python's
    `...00Z` and the backend/TS `...00.000Z` must collapse to one key.
    """
    try:
        normalized = period_start[:-1] + "+00:00" if period_start.endswith("Z") else period_start
        parsed = datetime.fromisoformat(normalized)
    except ValueError:
        return period_start
    if parsed.tzinfo is None:
        return period_start

    utc = parsed.astimezone(timezone.utc)
    milliseconds = utc.microsecond // 1000
    utc = utc.replace(microsecond=milliseconds * 1000)
    return f"{utc:%Y-%m-%dT%H:%M:%S}.{milliseconds:03d}Z"


def _key_of(k: AccumulatorKey) -> str | None:
    if k.scope == "pooled":
        return f"{k.rule_id}:__pooled__:{_canonical_period_start(k.period_start)}"
    # per_customer with no customer identity has no valid key. Callers treat
    # None as "skip" — collapsing to the pooled token here made per-customer
    # hard stops silently unenforceable for calls without a customer context
    # (events attribute those to 'anonymous', so the tokens never matched).
    if not k.customer_id:
        return None
    return f"{k.rule_id}:{k.customer_id}:{_canonical_period_start(k.period_start)}"


def _touch(key: str, entry: AccumulatorEntry) -> None:
    _accum.move_to_end(key)
    entry.last_touched = time.time()


def _maybe_evict() -> None:
    global _lru_warned_at
    if len(_accum) <= LRU_MAX:
        return
    overflow = len(_accum) - LRU_MAX
    for _ in range(overflow):
        _accum.popitem(last=False)
    now = time.time()
    if now - _lru_warned_at > 5 * 60:
        _lru_warned_at = now
        print(
            f"[pylva] budget accumulator LRU-evicted {overflow} entries (cap {LRU_MAX})",
            flush=True,
        )


def _ensure(key: str) -> AccumulatorEntry:
    entry = _accum.get(key)
    if entry is None:
        entry = AccumulatorEntry(total_usd=0.0, event_count=0, last_touched=time.time())
        _accum[key] = entry
        _maybe_evict()
    return entry


def get(k: AccumulatorKey) -> AccumulatorEntry:
    """Read the accumulator entry for a key (or a zeroed one if absent)."""
    key = _key_of(k)
    if key is None:
        return AccumulatorEntry()
    with _lock:
        entry = _accum.get(key)
        return entry if entry is not None else AccumulatorEntry()


def add(
    k: AccumulatorKey,
    actual_usd: float,
    *,
    expected_config_generation: int | None = None,
) -> None:
    """Bump the accumulator after a successful LLM call."""
    if actual_usd is None or actual_usd < 0:
        return
    key = _key_of(k)
    if key is None:
        return
    current_config_generation = get_config_generation()
    with _lock:
        if expected_config_generation is not None and (
            current_config_generation != expected_config_generation
            or not _local_context_is_current_locked(_cache_epoch, expected_config_generation)
        ):
            return
        entry = _ensure(key)
        entry.total_usd += actual_usd
        entry.event_count += 1
        _touch(key, entry)


def mark_exceeded_from_backend(
    *,
    rule_id: str,
    customer_id: str | None,
    limit_usd: float,
    period_start: str,
    expected_config_generation: int | None = None,
) -> None:
    """Backend ingest flagged this key as exceeded — bump local to limit+1."""
    k = AccumulatorKey(
        rule_id=rule_id,
        scope="pooled" if customer_id is None else "per_customer",
        customer_id=customer_id,
        period_start=period_start,
    )
    key = _key_of(k)
    if key is None:
        return
    current_config_generation = get_config_generation()
    with _lock:
        if expected_config_generation is not None and (
            current_config_generation != expected_config_generation
            or not _local_context_is_current_locked(_cache_epoch, expected_config_generation)
        ):
            return
        entry = _ensure(key)
        entry.total_usd = max(entry.total_usd, limit_usd + 1)
        entry.exceeded_source = "backend_ingest_flag"
        _touch(key, entry)


def set_from_sync(k: AccumulatorKey, server_total_usd: float) -> None:
    """Replace local total with the server truth (I-T3-3 overwrite semantics)."""
    if server_total_usd is None or server_total_usd < 0:
        return
    key = _key_of(k)
    if key is None:
        return
    with _lock:
        entry = _ensure(key)
        entry.total_usd = server_total_usd
        entry.exceeded_source = None
        _touch(key, entry)


def _set_from_sync_if_current(
    k: AccumulatorKey,
    server_total_usd: float,
    *,
    expected_epoch: int,
    expected_config_generation: int,
) -> None:
    """Apply reconciliation only while it still belongs to the active builder."""
    if server_total_usd is None or server_total_usd < 0:
        return
    key = _key_of(k)
    if key is None:
        return
    current_config_generation = get_config_generation()
    with _lock:
        if (
            current_config_generation != expected_config_generation
            or not _local_context_is_current_locked(expected_epoch, expected_config_generation)
        ):
            return
        entry = _ensure(key)
        entry.total_usd = server_total_usd
        entry.exceeded_source = None
        _touch(key, entry)


@dataclass(frozen=True)
class PreCallCheckResult:
    over_limit: bool
    accumulated_usd: float
    projected_usd: float
    source: Literal["backend_ingest_flag"] | None = None


def check(
    *,
    rule_id: str,
    scope: Scope,
    customer_id: str | None,
    period_start: str,
    estimated_usd: float,
    limit_usd: float,
) -> PreCallCheckResult:
    """Non-throwing pre-call check. Wrapper decides whether to throw or warn."""
    key = _key_of(AccumulatorKey(rule_id, scope, customer_id, period_start))
    entry: AccumulatorEntry | None = None
    if key is not None:
        with _lock:
            entry = _accum.get(key)
            # LRU-touch on read: an actively-blocking key must not be evicted
            # just because nothing writes to it anymore (writes stop once
            # calls block).
            if entry is not None:
                _touch(key, entry)
    accumulated = entry.total_usd if entry is not None else 0.0
    projected = accumulated + (estimated_usd if estimated_usd > 0 else 0.0)
    return PreCallCheckResult(
        # >= matches the server (computeBudgetExceededFlags / budget sync both
        # flag at total >= limit) so spend exactly at the limit blocks on both
        # sides of the contract.
        over_limit=projected >= limit_usd,
        accumulated_usd=accumulated,
        projected_usd=projected,
        source=entry.exceeded_source if entry is not None else None,
    )


def start_sync_loop() -> None:
    """Start the 5-minute sync loop. Idempotent."""
    global _sync_timer
    config_generation = get_config_generation()
    with _lock:
        loop_epoch = _cache_epoch
        if not _local_context_is_current_locked(loop_epoch, config_generation):
            return

    def _schedule_next() -> None:
        global _sync_timer
        current_config_generation = get_config_generation()
        with _lock:
            current = (
                current_config_generation == config_generation
                and _local_context_is_current_locked(loop_epoch, config_generation)
            )
        if not current:
            return
        with _timer_lock:
            with _lock:
                if (
                    current_config_generation != config_generation
                    or not _local_context_is_current_locked(loop_epoch, config_generation)
                ):
                    return
            if _sync_timer is not None:
                return

            timer: threading.Timer

            def tick() -> None:
                global _sync_timer
                with _timer_lock:
                    if _sync_timer is not timer:
                        return
                    _sync_timer = None
                try:
                    _run_sync_now(loop_epoch, config_generation)
                except Exception:  # R1 — never crash the host process
                    pass
                finally:
                    _schedule_next()

            timer = threading.Timer(SYNC_INTERVAL_SEC, tick)
            timer.daemon = True
            _sync_timer = timer
            timer.start()

    _schedule_next()


def stop_sync_loop() -> None:
    global _sync_timer
    with _timer_lock:
        if _sync_timer is not None:
            _sync_timer.cancel()
            _sync_timer = None


def _warn_missing_period_start_fallback() -> None:
    global _sync_missing_period_start_warned
    if _sync_missing_period_start_warned:
        return
    _sync_missing_period_start_warned = True
    print(
        "[pylva] /budget/sync response is missing period_start; "
        "falling back to an unambiguous legacy match",
        flush=True,
    )


def _warn_ambiguous_period_start_skip() -> None:
    global _sync_ambiguous_period_start_warned
    if _sync_ambiguous_period_start_warned:
        return
    _sync_ambiguous_period_start_warned = True
    print(
        "[pylva] /budget/sync response is missing period_start for "
        "multiple local periods; skipping ambiguous reconciliation",
        flush=True,
    )


def _matching_sync_snapshot(
    snapshot: list[dict[str, object]],
    response_entry: dict[str, object],
) -> dict[str, object] | None:
    tuple_matches = [
        s
        for s in snapshot
        if s["rule_id"] == response_entry.get("rule_id")
        and s["scope"] == response_entry.get("scope")
        and s["customer_id"] == response_entry.get("customer_id")
    ]
    response_period_start = response_entry.get("period_start")
    if isinstance(response_period_start, str):
        canonical_response_period_start = _canonical_period_start(response_period_start)
        return next(
            (s for s in tuple_matches if s["period_start"] == canonical_response_period_start),
            None,
        )
    if len(tuple_matches) == 1:
        _warn_missing_period_start_fallback()
        return tuple_matches[0]
    if len(tuple_matches) > 1:
        _warn_ambiguous_period_start_skip()
    return None


def run_sync_now() -> None:
    """POST current accumulator state to /api/v1/budget/sync, overwrite local."""
    expected_config_generation = get_config_generation()
    with _lock:
        expected_epoch = _cache_epoch
    _run_sync_now(expected_epoch, expected_config_generation)


def _run_sync_now(expected_epoch: int, expected_config_generation: int) -> None:
    cfg = get_config()
    if cfg is None:
        return
    current_config_generation = get_config_generation()
    with _lock:
        if (
            current_config_generation != expected_config_generation
            or not _local_context_is_current_locked(expected_epoch, expected_config_generation)
        ):
            return
        if not _accum:
            return
        snapshot: list[dict[str, object]] = []
        for composite, entry in _accum.items():
            rule_id, scope_token, period_start = composite.split(":", 2)
            scope: Scope = "pooled" if scope_token == "__pooled__" else "per_customer"
            customer_id = None if scope == "pooled" else scope_token
            snapshot.append(
                {
                    "rule_id": rule_id,
                    "scope": scope,
                    "customer_id": customer_id,
                    "accumulated_cost_usd": entry.total_usd,
                    "period_start": period_start,
                    "event_count": entry.event_count,
                }
            )

    try:
        with httpx.Client(timeout=10.0) as client:
            for offset in range(0, len(snapshot), SYNC_BATCH_SIZE):
                batch = snapshot[offset : offset + SYNC_BATCH_SIZE]
                try:
                    resp = client.post(
                        f"{cfg.endpoint}/api/v1/budget/sync",
                        headers={"X-Pylva-Key": cfg.api_key, "Content-Type": "application/json"},
                        json={"entries": batch},
                    )
                    if not resp.is_success:
                        continue
                    body = resp.json()
                    entries = body.get("entries", [])
                except httpx.RequestError:
                    # A transport failure applies to the endpoint, not one
                    # malformed batch. Stop so synchronous init cannot turn
                    # one outage into up to 100 sequential timeout waits.
                    return

                for r in entries:
                    snap = _matching_sync_snapshot(batch, r)
                    if snap is None:
                        continue
                    snap_period_start = snap["period_start"]
                    if not isinstance(snap_period_start, str):
                        continue
                    _set_from_sync_if_current(
                        AccumulatorKey(
                            rule_id=r["rule_id"],
                            scope=r["scope"],
                            customer_id=r["customer_id"],
                            period_start=snap_period_start,
                        ),
                        r["server_total_usd"],
                        expected_epoch=expected_epoch,
                        expected_config_generation=expected_config_generation,
                    )
    except httpx.RequestError:
        return


def init_accumulator() -> None:
    """Prime the accumulator on SDK init. Non-blocking enough for a host that
    imports pylva and immediately makes calls — first call falls through
    passthrough if backend is unreachable."""
    start_sync_loop()
    try:
        run_sync_now()
    except Exception:
        pass


def _local_context_is_current_locked(expected_epoch: int, expected_config_generation: int) -> bool:
    return expected_epoch == _cache_epoch and (
        _accepted_config_generation is None
        or expected_config_generation == _accepted_config_generation
    )


def _invalidate_accumulator_for_config_change(
    next_config_generation: int | None = None,
) -> None:
    """Drop tenant-scoped totals and stop the old builder's reconciliation loop.

    ``next_config_generation`` lets config pause sync work when this hook is
    invoked immediately before installing the replacement identity.
    """
    global _accepted_config_generation, _cache_epoch
    global _lru_warned_at, _sync_missing_period_start_warned
    global _sync_ambiguous_period_start_warned
    with _lock:
        _cache_epoch += 1
        _accepted_config_generation = next_config_generation
        _accum.clear()
        _lru_warned_at = 0.0
        _sync_missing_period_start_warned = False
        _sync_ambiguous_period_start_warned = False
    stop_sync_loop()


def _reset_accumulator_for_tests() -> None:
    global _accepted_config_generation, _cache_epoch, _lru_warned_at
    global _sync_missing_period_start_warned, _sync_ambiguous_period_start_warned
    with _lock:
        _cache_epoch += 1
        _accepted_config_generation = None
        _accum.clear()
        _lru_warned_at = 0.0
        _sync_missing_period_start_warned = False
        _sync_ambiguous_period_start_warned = False
    stop_sync_loop()

File: packages/sdk-ts/src/core/budget_accumulator.ts
// B2a — SDK-side budget accumulator (D3 / §6).
//
// Per-SDK-process Map keyed on `${rule_id}:${scope_token}:${period_start}`
// where scope_token = scope === 'pooled' ? '__pooled__' : customer_id.
// This is a *near-real-time* signal on long-lived ECS tasks; cross-container
// drift is bounded by the ingest round-trip × concurrent SDK instances. The
// backend ingest-response `budget_exceeded[]` flag is the authoritative stop
// (I-T3-2).
//
// Eviction: LRU at 50,000 entries. Overflow is logged once per 5-min window.
// Sync: onInit POSTs current state to /api/v1/budget/sync to prime; a loop
// every SYNC_INTERVAL_MS re-reconciles. server_total REPLACES local (I-T3-3).

import { getConfig } from './config.js';
import type { BudgetExceededFlag } from '@pylva/shared/budget-errors';
import { registerIdentityResetter } from './identity_registry.js';
import { AuthenticatedRoute, coreRuntime } from '../internal/core-runtime-state.js';

const LRU_MAX = 50_000;
const SYNC_INTERVAL_MS = 5 * 60 * 1000;
// The backend caps one /budget/sync request at 500 entries. Large builders can
// accumulate many more per-customer keys, so reconcile in bounded batches
// instead of letting the server reject the entire snapshot.
const SYNC_BATCH_SIZE = 500;

export type Scope = 'per_customer' | 'pooled';
export type Period = 'hour' | 'day' | 'week' | 'month';

interface AccumulatorEntry {
  total_usd: number;
  event_count: number;
  last_touched: number;
  exceeded_source?: 'backend_ingest_flag';
}

interface AccumulatorKey {
  rule_id: string;
  scope: Scope;
  customer_id: string | null;
  period_start: string;
}

// Map preserves insertion order → we use that for LRU-eviction.
const accumulator = new Map<string, AccumulatorEntry>();

let syncTimer: ReturnType<typeof setInterval> | null = null;
let lruWarnedAt = 0;
let syncInFlight: Promise<void> | null = null;
let syncMissingPeriodStartWarned = false;
let syncAmbiguousPeriodStartWarned = false;
let accumulatorEpoch = 0;
const activeControllers = new Set<AbortController>();

function keyOf(k: AccumulatorKey): string | null {
  if (k.scope === 'pooled') return `${k.rule_id}:__pooled__:${k.period_start}`;
  // per_customer with no customer identity has no valid key. Callers treat
  // null as "skip" — collapsing to the pooled token here made per-customer
  // hard stops silently unenforceable for calls without a customer context
  // (events attribute those to 'anonymous', so the tokens never matched).
  if (k.customer_id === null || k.customer_id === '') return null;
  return `${k.rule_id}:${k.customer_id}:${k.period_start}`;
}

function touchEntry(key: string, entry: AccumulatorEntry): void {
  // LRU touch: delete then re-insert to move to the end.
  accumulator.delete(key);
  entry.last_touched = Date.now();
  accumulator.set(key, entry);
}

function maybeEvict(): void {
  if (accumulator.size <= LRU_MAX) return;
  const toEvict = accumulator.size - LRU_MAX;
  const iter = accumulator.keys();
  for (let i = 0; i < toEvict; i++) {
    const oldest = iter.next().value;
    if (oldest !== undefined) accumulator.delete(oldest);
  }
  const now = Date.now();
  if (now - lruWarnedAt > 5 * 60 * 1000) {
    lruWarnedAt = now;
    console.warn(`[pylva] budget accumulator LRU-evicted ${toEvict} entries (size cap ${LRU_MAX})`);
  }
}

/**
 * Fetch-or-create the entry for a key. Does not mutate total; use `add` or
 * `setFromSync` for that.
 */
function ensureEntry(key: string): AccumulatorEntry {
  let entry = accumulator.get(key);
  if (!entry) {
    entry = { total_usd: 0, event_count: 0, last_touched: Date.now() };
    accumulator.set(key, entry);
    maybeEvict();
  }
  return entry;
}

/** Read accumulator state for a key. Returns a zeroed entry if absent. */
export function get(k: AccumulatorKey): Readonly<AccumulatorEntry> {
  const key = keyOf(k);
  const entry = key === null ? undefined : accumulator.get(key);
  return entry ?? { total_usd: 0, event_count: 0, last_touched: 0 };
}

/**
 * Bump the accumulator after a call completes. Call with the ACTUAL post-call
 * cost (see core/budget_rules recordLlmSpend).
 */
export function add(k: AccumulatorKey, actual_usd: number): void {
  if (!Number.isFinite(actual_usd) || actual_usd < 0) return;
  const key = keyOf(k);
  if (key === null) return;
  const entry = ensureEntry(key);
  entry.total_usd += actual_usd;
  entry.event_count += 1;
  touchEntry(key, entry);
}

/**
 * Backend-authoritative override: called when /api/v1/events returns a
 * budget_exceeded flag. Bumps local accumulator to `limit_usd + 1` so the
 * next pre-call check for that key throws (I-T3-2, §6.4 edge row
 * "`budget_exceeded` from ingest").
 */
export function markExceededFromBackend(flag: BudgetExceededFlag): void {
  const k: AccumulatorKey = {
    rule_id: flag.rule_id,
    scope: flag.customer_id === null ? 'pooled' : 'per_customer',
    customer_id: flag.customer_id,
    period_start: flag.period_start,
  };
  const key = keyOf(k);
  if (key === null) return;
  const entry = ensureEntry(key);
  entry.total_usd = Math.max(entry.total_usd, flag.limit_usd + 1);
  entry.exceeded_source = 'backend_ingest_flag';
  touchEntry(key, entry);
}

/**
 * Replace (not add) the accumulator total for a key — used by `/budget/sync`
 * reconciliation. Spec I-T3-3: server_total_usd REPLACES local, never adds.
 */
export function setFromSync(k: AccumulatorKey, server_total_usd: number): void {
  if (!Number.isFinite(server_total_usd) || server_total_usd < 0) return;
  const key = keyOf(k);
  if (key === null) return;
  const entry = ensureEntry(key);
  entry.total_usd = server_total_usd;
  entry.exceeded_source = undefined;
  touchEntry(key, entry);
}

export interface PreCallCheckInput extends AccumulatorKey {
  estimated_usd: number;
  limit_usd: number;
}

export interface PreCallCheckResult {
  over_limit: boolean;
  accumulated_usd: number;
  projected_usd: number; // accumulated + estimated
  source?: 'backend_ingest_flag';
}

/**
 * Non-throwing pre-call check: returns whether the projected spend (current
 * accumulator + estimated next-call cost) crosses the budget. The wrapper
 * uses this and decides whether to throw (hard_stop) or emit an advisory
 * warning (soft budget).
 */
export function check(input: PreCallCheckInput): PreCallCheckResult {
  const key = keyOf(input);
  const entry = key === null ? undefined : accumulator.get(key);
  // LRU-touch on read: an actively-blocking key must not be evicted just
  // because nothing writes to it anymore (writes stop once calls block).
  if (key !== null && entry) touchEntry(key, entry);
  const accumulated_usd = entry?.total_usd ?? 0;
  const projected_usd = accumulated_usd + (input.estimated_usd > 0 ? input.estimated_usd : 0);
  return {
    // >= matches the server (computeBudgetExceededFlags / reconcileBudgetSync
    // both flag at total >= limit) so spend exactly at the limit blocks on
    // both sides of the contract.
    over_limit: projected_usd >= input.limit_usd,
    accumulated_usd,
    projected_usd,
    ...(entry?.exceeded_source ? { source: entry.exceeded_source } : {}),
  };
}

/**
 * Start the 5-minute sync loop. Safe to call multiple times — subsequent
 * calls are no-ops.
 */
export function startSyncLoop(): void {
  if (syncTimer) return;
  // First sync runs after SYNC_INTERVAL_MS; for init-time priming the SDK
  // calls `runSyncNow` explicitly from `initAccumulator`.
  syncTimer = setInterval(() => {
    void runSyncNow().catch(() => {
      /* R1 — sync failure must never crash host */
    });
  }, SYNC_INTERVAL_MS);
  // Node-only: unref so the loop doesn't keep the process alive.
  if (typeof (syncTimer as unknown as { unref?: () => void }).unref === 'function') {
    (syncTimer as unknown as { unref: () => void }).unref();
  }
}

export function stopSyncLoop(): void {
  if (syncTimer) {
    clearInterval(syncTimer);
    syncTimer = null;
  }
}

interface BudgetSyncSnapshotEntry {
  rule_id: string;
  scope: Scope;
  customer_id: string | null;
  accumulated_cost_usd: number;
  period_start: string;
  event_count: number;
}

interface BudgetSyncResponseEntry {
  rule_id: string;
  scope: Scope;
  customer_id: string | null;
  period_start?: string;
  server_total_usd: number;
}

function warnMissingPeriodStartFallback(): void {
  if (syncMissingPeriodStartWarned) return;
  syncMissingPeriodStartWarned = true;
  console.warn(
    '[pylva] /budget/sync response is missing period_start; falling back to an unambiguous legacy match',
  );
}

function warnAmbiguousPeriodStartSkip(): void {
  if (syncAmbiguousPeriodStartWarned) return;
  syncAmbiguousPeriodStartWarned = true;
  console.warn(
    '[pylva] /budget/sync response is missing period_start for multiple local periods; skipping ambiguous reconciliation',
  );
}

function findSnapshotForSyncEntry(
  snapshot: BudgetSyncSnapshotEntry[],
  responseEntry: BudgetSyncResponseEntry,
): BudgetSyncSnapshotEntry | null {
  const tupleMatches = snapshot.filter(
    (s) =>
      s.rule_id === responseEntry.rule_id &&
      s.scope === responseEntry.scope &&
      s.customer_id === responseEntry.customer_id,
  );
  if (typeof responseEntry.period_start === 'string') {
    return tupleMatches.find((s) => s.period_start === responseEntry.period_start) ?? null;
  }
  if (tupleMatches.length === 1) {
    warnMissingPeriodStartFallback();
    return tupleMatches[0]!;
  }
  if (tupleMatches.length > 1) warnAmbiguousPeriodStartSkip();
  return null;
}

/**
 * POST the current accumulator snapshot to /api/v1/budget/sync and overwrite
 * local entries with the server truth. Coalesced via `syncInFlight` so
 * concurrent callers share one request.
 */
export async function runSyncNow(): Promise<void> {
  if (syncInFlight) return syncInFlight;
  const owner = accumulatorEpoch;
  const promise = doSync(owner);
  const wrapped = promise.finally(() => {
    if (syncInFlight === wrapped) syncInFlight = null;
  });
  syncInFlight = wrapped;
  return syncInFlight;
}

async function doSync(owner: number): Promise<void> {
  if (!getConfig()) return;
  if (accumulator.size === 0) return;

  const snapshot: BudgetSyncSnapshotEntry[] = [...accumulator.entries()].map(
    ([compositeKey, entry]) => {
      // Key shape is `${rule_id}:${scope_token}:${period_start}`. period_start
      // is an ISO-8601 timestamp that itself contains colons (e.g.
      // `2026-06-09T00:00:00.000Z`), so we split on the FIRST two colons only and
      // keep the remainder intact. A bare `split(':')` with array destructuring
      // truncates period_start at the first inner colon (`2026-06-09T00`),
      // corrupting the value POSTed to /budget/sync and making setFromSync write
      // to a phantom key that never reconciles the real accumulator entry
      // (I-T3-3). Mirrors the Python SDK's `split(':', 2)`.
      const parts = compositeKey.split(':');
      const rule_id = parts[0] ?? '';
      const scope_token = parts[1] ?? '';
      const period_start = parts.slice(2).join(':');
      const scope: Scope = scope_token === '__pooled__' ? 'pooled' : 'per_customer';
      const customer_id = scope === 'pooled' ? null : scope_token;
      return {
        rule_id,
        scope,
        customer_id,
        accumulated_cost_usd: entry.total_usd,
        period_start,
        event_count: entry.event_count,
      };
    },
  );

  for (let offset = 0; offset < snapshot.length; offset += SYNC_BATCH_SIZE) {
    if (owner !== accumulatorEpoch) return;
    const batch = snapshot.slice(offset, offset + SYNC_BATCH_SIZE);
    const controller = new AbortController();
    activeControllers.add(controller);
    try {
      const res = await coreRuntime.authenticatedRequest({
        route: AuthenticatedRoute.BUDGET_SYNC,
        body: JSON.stringify({ entries: batch }),
        signal: controller.signal,
      });
      if (owner !== accumulatorEpoch) return;
      if (!res.ok) continue; // R5 passthrough — reconcile other batches if possible
      const body = JSON.parse(res.bodyText) as {
        entries?: BudgetSyncResponseEntry[];
      };
      if (owner !== accumulatorEpoch) return;
      const entries = body.entries ?? [];
      for (const r of entries) {
        if (owner !== accumulatorEpoch) return;
        const snap = findSnapshotForSyncEntry(batch, r);
        if (!snap) continue;
        setFromSync(
          {
            rule_id: r.rule_id,
            scope: r.scope,
            customer_id: r.customer_id,
            period_start: snap.period_start,
          },
          r.server_total_usd,
        );
      }
    } catch {
      // A transport failure applies to the endpoint, not one malformed batch.
      // Stop here so a large snapshot cannot turn one outage into up to 100
      // sequential retries (R1/R5). The next scheduled sync retries everything.
      return;
    } finally {
      activeControllers.delete(controller);
    }
  }
}

/**
 * Prime the accumulator on SDK init. Non-blocking: kicked off by `init()` via
 * `void initAccumulator()`. Safe under cold-boot + backend-down (fresh-boot
 * passthrough — I-T3-1).
 */
export async function initAccumulator(): Promise<void> {
  startSyncLoop();
  await runSyncNow();
}

// Test helper: drains local state so tests don't leak between runs.
export function _resetAccumulatorForTests(): void {
  resetAccumulator();
}

function resetAccumulator(): void {
  accumulatorEpoch += 1;
  for (const controller of activeControllers) controller.abort();
  activeControllers.clear();
  accumulator.clear();
  lruWarnedAt = 0;
  syncMissingPeriodStartWarned = false;
  syncAmbiguousPeriodStartWarned = false;
  syncInFlight = null;
  stopSyncLoop();
}

export function _resetAccumulatorForIdentityChange(): void {
  resetAccumulator();
}

registerIdentityResetter(_resetAccumulatorForIdentityChange);
